### ISSUE TITLE
Sample indexes

### ISSUE THREAD

**Issue by vturrisi** [libffcv/ffcv#87]
Is there any way to access specific sample indexes, e.g., given a minibatch, is it possible to know the unique index of each sample in it?

**Comment by andrewilyas** [comment]
@vturrisi Yes! Inside a transform, you can set the with_indices flag to get access to image indices (https://docs.ffcv.io/ffcv_examples/transform_with_inds.html). You can also pass an ``indices=`` flag to the Loader in order to specify which indices you want to read (e.g., if you want to read indices in a specific order, you can pass a permutation of np.arange(dataset_len) to ``indices`` and then use OrderOption.Sequential), lmk if this helps!

**Comment by vturrisi** [comment]
Hmm, I think that's not exactly what I need. What I want to do is access the indexes outside of the pipeline, something like
```
train_loader = Loader(....)
for idx, imgs, label in train_loader:
    ....
```
I don't need to influence the order the images are read, just need to access this info from outside. Is it doable?

**Comment by andrewilyas** [comment]
I don't think it should be too hard to add support for this, let me think about whether its possible to do natively. In the meantime, the easiest possible workaround is just writing the beton file with indices (this will add no overhead at all when you don't want indices as you can just set ``pipelines={'idx': None, ...}`` when constructing the loader and it won't load the indices at all). 

If re-writing the baton file is too expensive (I don't think it should be, since the writing is usually pretty fast even for imagenet), I think there's a way to do this with a custom transform (but will have to think a bit more).

**Comment by GuillaumeLeclerc** [comment]
We could have a `with_indices` on the `Loader` that would be very easy to add

**Comment by GuillaumeLeclerc** [comment]
I'm going to start a new a `v0.0.3` branch and start adding all these little features and people can start testing before we release it.

**Comment by GuillaumeLeclerc** [comment]
I was adding this when I realized that just having a `with_indices` it might be problematic as there will be no way to copy them to the GPU or do anything with it. Is it something we think will be useful or should we stay with `with_indices` @andrewilyas  @vturrisi ?

**Comment by vturrisi** [comment]
@GuillaumeLeclerc sorry, I didn't get it. If `with_indices` is problematic, what's the other option?

**Comment by andrewilyas** [comment]
@vturrisi I think what Guillaume is saying is that with_indices would just return the indices as a scalar array (compare this to, e.g., an ``index`` pipeline that could do augmentations, do asynchronous copying to the GPU, etc). Maybe we could have an ``index_pipeline`` argument instead of ``with_indices`` that could fulfill this purpose without having to rewrite the baton file?

**Comment by GuillaumeLeclerc** [comment]
Is it something we want in the next release ? Anyone down to write the unit tests/documentation for this. I have a bit too much on my plate right now :/

**Comment by vturrisi** [comment]
Personally speaking, I don't think it's something urgent. I'm waiting for implementations of common transformations (like color jitter), or for me to have time to implement them myself, to switch to fccv.
